UI: Filter Revision - Rules and Interface

tfamula opened the pull request:
Hi @Amstutz & @klees,
@ezenzen and me did the first steps for the Filter Revision. This PR includes:

- a big revision for the description and rules of the Filter to make clear the new/changed appearance and behavior of the component
- the introduction of a new method `isComplex` in the `FilterInput` Interface to differentiate between Input Fields which can be edited directly and Input Fields which need to be rendered in a Popover

At the moment, I have not spotted other necessary changes for the public interface, which have to be discussed in the Jour Fixe.

We are looking forward to your feedback!


Greetings

tfamula committed:
Filter Revision: Rules and Interface

tfamula commented:
BTW: Due to the amount of text in the "composition" and "effect" sections, these segments are hard to read in the Kitchen Sink Documentation. Unfortunately, it is not possible to do some formatting (or I don't know how), e.g. with paragraphs and bullet points. Is it maybe planned to implement this in the future?

Amstutz commented:
Thx a lot to everybody involved. This is a big step forwards for the filter in the UI Components, thx.

@klees and @amstutz discussed the proposal of the interface today. We also asked @yvseiler for an opinion on the final wording on the concept. She might have some minor feedback on parts on the content. If so, she will propose those them until monday. 

We like the isComplex change on the interface, since it promotes the interface from a pure marker to an interface with some meaningful semantics. Note: "isComplex" might even be an aspect that could be interesting for input fields independently from filters in the long run.

We are looking forward to discuss the change in the upcoming JF.

yvseiler reviewed (commented):
Hi @tfamula 

It is great to see how the filter concept already discussed earlier by the JF is now being implemented. Thank you very much for your cooperation and thinking through the whole revision! The filters will thus receive a great upgrade in use.

Please consider the following suggestions:
- [x] **Resolve duplications**: To shorten the text a little more, I suggest deleting the two sentences [here](https://github.com/ILIAS-eLearning/ILIAS/pull/3333/files#diff-132413157cba32aac68b271318a8ddc8affc411ef496624b745aa3ca29da5351R56-R58); "Only the Input Fields... are not visible in this area." because this is already introduced in the sentence line 53.
- [ ] **Always clickable buttons**: It is written [here](https://github.com/ILIAS-eLearning/ILIAS/pull/3333/files#diff-132413157cba32aac68b271318a8ddc8affc411ef496624b745aa3ca29da5351R115-R117) that the buttons (Apply and Reset) should be clickable, whether it has an effect on the filter or the item collection or not. The consequence of this is that it is not made visible to a user whether his new filter values are already filtered or not. What was the reason for deciding that the buttons should be clickable in any case? Possibly certain decisions led to this decision, which I cannot (any more) reproduce at the moment. I think it would be good to think about whether it might not make sense to show the user whether his last changes to the filter have already been applied or not (f.e. in which the buttons are only clickable if they also result in a change).

Thank you a lot for contributing to ILIAS.

Best regards,
@yvseiler

yvseiler commented on src/UI/Component/Input/Container/Factory.php:
```diff
@@ -39,57 +39,122 @@ public function form();
+     *      In the Filter Content Area, the Apply Button and Reset Button can be used. Clicking on the Apply Button
+     *      has multiple effects: The appearance of the Input Fields, i.e. if an Input Field is shown or hidden in the
+     *      Filter Content Area, will be saved. The values, which are entered into the Input Fields, are applied. The
+     *      content of the item colletion is reloaded immediately according to the Filter settings. The status of the
```
```suggestion
     *      content of the item collection is reloaded immediately according to the Filter settings. The status of the
```

matthiaskunkel commented:
Jour Fixe, 03 MAY 2021 : We highly appreciate the suggested rules and interface and accept the PR for trunk.

tfamula committed:
Adjust Filter rules and implement isComplex for Filter Inputs

tfamula commented:
Hello all!

I adjusted the rules according to Yvonne's comments.
I also implemented the `isComplex` method for the Filter Inputs, so that the Unit Tests will not fail anymore (It seems the Tag Input was accidentally a Filter Input, because there is no rendering for it in the `FilterContextRenderer` and you get an error when using it in the Filter).

As we already mentioned in the Jour Fixe, the concept of the "not always clickable buttons" will not be considered in this step of the Filter Revision. It needs an extra conceptual discussion and should definitively be implemented/discussed in a separate PR. IMO, the concept can also be liable for bugs/issues. For now, we do not see real disadvantages when the Apply Button and Reset Button are always clickable, since this is usual behavior for filters on other platforms, too.

I would appreciate it if this PR could be merged. 
Next week, I will create a separate PR for the implementation part of the the improved Filter.


Greetings
Thomas

Amstutz commented:
Thx a lot for this first part towards better filters! Thx to everybody involved. Looking forward to the implementation.

tfamula commented:
Implementation part: #3390
